### ISSUE TITLE
Book: PDF output with SVG images

### ISSUE THREAD

**Issue by pat-s** [quarto-dev/quarto-cli#997]
I've read #196 and the linked documentation but still do not know how to proceed.

We're currently aiming to switch from a large bookdown book to quarto.

For the PDF output we get errors like

```tex
compilation failed- error
LaTeX Error: Cannot determine size of graphic in ./images/learner.svg (no Bou
ndingBox).

See the LaTeX manual or LaTeX Companion for explanation.
Type  H <return>  for immediate help.
 ...                                              
                                                  
l.1636 \includegraphics{./images/learner.svg}
```

The source references the file via

````r
```{r 02-basics-001, echo = FALSE}
knitr::include_graphics("images/ml_abstraction.svg")
```
````

This works with bookdown but quarto does not seem to take it directly.

Can quarto already handle such cases dynamically?

**Comment by jjallaire** [comment]
@cderv Does bookdown do something automatic to handle SVG files in LaTeX output?

**Comment by cderv** [comment]
I don't think we are doing anything specific. SVG in LaTeX does not play well. A conversion from SVG images to PDF images is usually needed.  It could be done automatically somehow using a tool like `rsvg-convert` or just R package **magick**. This was discussed in https://github.com/rstudio/bookdown/issues/658. We did not include anything specific in **bookdown** or **knitr** and let a use do the conversion itself. 

So usually, for HTML + PDF project, what happens is a `PNG` or `SVG` version of the image is provided for HTML output, **and** a PDF version for the LaTeX version. `knitr::include_graphics()` has a special argument `auto_pdf` to set to `TRUE` so that the image extension will be changed to `.pdf` if it exists. This is not default behavior, either `auto_pdf = TRUE` in each call to `include_graphics()` either there is an option `knitr.graphics.auto_pdf` you can set to TRUE. 
With this, this means that in 
````r
knitr::include_graphics("images/ml_abstraction.svg")
````
it will be rendered by **knitr** as 
````tex
\includegraphics{./images/learner.pdf}
````

@pat-s I believe you don't have this feature activated in your new quarto version of the project. 
It is working in **bookdown** for you because: 

* You have pdf version of the svg image: https://github.com/mlr-org/mlr3book/blob/main/bookdown/images/learner.pdf
* You are using `include_graphics()` to include image
* You are setting the **knitr** option in your `setup.R` script: https://github.com/mlr-org/mlr3book/blob/main/bookdown/setup.R#L7
* This `setup.R` script is loaded in **bookdown** using `before_chapter_script` : https://github.com/mlr-org/mlr3book/blob/main/bookdown/_bookdown.yml#L4

I believe this is why it is working in your **bookdown** version. 

So how to fix that in your Quarto project ? 

Quarto does not have a `before_chapter_script` option. One way is to use a setup chunk in each file like this: 
````markdown
```{r, file = "setup.R", include = FALSE}
```
````

Using a `.Rprofile` is another option, as it should be taken into account in the R process used during rendering each chapter.

Otherwise, depending on what needs to be set, one can also use special yaml option in Quarto doc. Like [knitr options in yaml](https://quarto.org/docs/computations/r.html#knitr-options) for `opts_knit` or `opts_chunk`. Setting R option should also be possible to set this way 
````yaml
knitr:
  opts_chunk: 
    R.options:
      knitr.graphics.auto_pdf: true
````

This could be set on a `.qmd` file YAML, globally in `_quarto.yml` or in a shared metadata file on folder basis [`_metadata.yaml`](https://quarto.org/docs/projects/quarto-projects.html#shared-metadata)

@pat-s I hope this help building the mlr3 book. 

@jjallaire if we want to support this better out of the box in Quarto, I think this would require doing a conversion in post processing step or within Lua filter when a SVG image is detected. 

Pandoc does it when converting to PDF I believe
https://github.com/jgm/pandoc/blob/ae8ed408b03688f4ddf6a13d0e55fba9138c6eb9/src/Text/Pandoc/PDF.hs#L202-L216
It is using rsvg-convert CLI tool, but there are several variation it seems, including a npm package: https://www.npmjs.com/package/rsvg

I am thinking this could also be printed to PDF probably as we are doing with Diagrams. wouldn't it ?

**Comment by pat-s** [comment]
@cderv Wow, again impressed by your level of detail when it comes to such replies. Just amazing.

Also you inferred that it's about the mlr3book and digged so deep - huge thanks!

I can happily report that I got it working 🥳 I went with the YAML option for defining `knitr.graphics.auto_pdf: true`.
I was not aware anymore that we had set this option via `before_chapter_script` in the bookdown setup. It might have even been me who did this but this is one point that is quickly overlooked.

Feel free to close if everythings has been said - I'll leave it open in case you want to continue discussin with JJ.

**Comment by jjallaire** [comment]
@cderv It does seem like we should consider doing this when rsvg-convert is available on the path. We'd just need a Lua filter for this that targets PDF/LaTeX/Beamer. I'd think we'd want to do it automatically (as the other alternative is an error message from LaTeX). Could you take a shot at a PR for this? (note that the node module requires native code so not something we can easily port to deno). 

cc @cscheid

**Comment by cderv** [comment]
>  I'd think we'd want to do it automatically (as the other alternative is an error message from LaTeX).

The other option I see is supporting natively in Quarto an option like `knitr.graphics.auto_pdf` that would change any image path extension to pdf for LaTeX outputs. This would allow a user to do like in mlr3 book and provide SVG for HTML and PDF for LaTeX outputs. When PNG are provided, it works ok in LaTeX and HTML. 

This solution would mean Quarto is not in charge of providing PDF version of images but let a user be responsible for providing several image format.

I'll give a shot though at the Lua Filter to see if we can provide an auto conversion feature.

**Comment by jjallaire** [comment]
Note that we already have support for multi-format figures (just provide the figure name without an extension and it will bind to the right thing): https://quarto.org/docs/authoring/figures.html#multiformat-figures

**Comment by cderv** [comment]
Oh perfect. I missed that. We may need to support path without extension in `knitr::include_graphics()` or maybe set `knitr.graphics.auto_pdf = TRUE` by default when **knitr** is used in a Quarto project.

**Comment by jjallaire** [comment]
I think if we could do this in Lua it would work everywhere (and then we wouldn't have to direct R users to `include_graphics`, they could just use images in markdown as they are accustomed to)

**Comment by fuhrmanator** [comment]
Here's a [MWE repo](https://github.com/fuhrmanator/quarto-test) of how I get SVG into PDF in a book with no magic (unless it's the latexmk that's doing the magic?). You can ignore the filters for the PlantUML/Graphviz diagrams - I made a sample `svg-test.qmd` that just links to an SVG file and it renders just fine in PDF. Hope it's somehow helpful.

**Comment by JER-ry** [comment]
Another way may be pop up a notice that the user should install Inkscape, add it to path, and then generate tex code like
/usepackage{svg}
/includesvg{fig}
Currently I do that manually but it should be automated right?
*From a Jupyter user who prefers to generate SVG

**Comment by SamEdwardes** [comment]
I came across this same issue today. I just want to add (although I think it is already implied) that this issue is not related to just the book format. I was unable to render the following project because I included an SVG image.

## Project

**_quarto.yml**

```yaml
project:
  type: default
```

**doc.qmd**

```md
---
title: "Hello world"
---

This is my quarto doc. Here is an SVG image:

![](example-image.drawio.svg)
```

**example-image.drawio.svg**

![example-image drawio](https://user-images.githubusercontent.com/18248949/190209144-d032c63d-e4d0-45df-8bda-93b03a468012.svg)

## Render

```bash
$ quarto render --to pdf
```

```log
pandoc 
  to: latex
  output-file: doc.tex
  standalone: true
  pdf-engine: xelatex
  variables:
    graphics: true
    tables: true
  default-image-extension: pdf
  
metadata
  documentclass: scrartcl
  classoption:
    - DIV=11
    - numbers=noendperiod
  papersize: letter
  header-includes:
    - '\KOMAoption{captions}{tableheading}'
  block-headings: true
  title: Hello world
  
running xelatex - 1
  This is XeTeX, Version 3.141592653-2.6-0.999994 (TeX Live 2022) (preloaded format=xelatex)
   restricted \write18 enabled.
  entering extended mode
  
updating tlmgr

updating existing packages

compilation failed- error
LaTeX Error: Cannot determine size of graphic in example-image.drawio.svg (no BoundingBox).

See the LaTeX manual or LaTeX Companion for explanation.
Type  H <return>  for immediate help.
 ...                                              
                                                  
l.155 ...ht=\textheight]{example-image.drawio.svg}
                                                   

see /Users/samedwardes/tmp/quarto-reprex/doc.log for more information.
```

**Comment by cscheid** [comment]
@SamEdwardes in your specific case, it will be hard for us to fix this automatically, since LaTeX's PDF engine doesn't take SVG files by default.

Nevertheless, you can convert that file to (eg) png yourself and then use [multiformat figures](https://quarto.org/docs/authoring/figures.html#figure-basics).

**Comment by SamEdwardes** [comment]
Very cool - thanks for explaining and sharing @cscheid!

**Comment by nluetts** [comment]
I am in the process of trying out quarto and switch from a pure markdown+pandoc based workflow.
Before trying quarto, I included svg in my markdown files w/o problems, pandoc would just convert to PDF fine, using xelatex as pdf-engine. If I convert my `qmd` file to markdown and than run pandoc on it, it also works fine, not too surprisingly.
I was surprised that `quarto render` would not work ...
So maybe this is just a problem of how pandoc/latex is setup in quarto?

**Comment by cderv** [comment]
@nluetts when you are using pandoc only, are you converting directly `--from markdown --to pdf` ? 

I believe the behavior you encounter is related to https://github.com/quarto-dev/quarto-cli/issues/1144

* Quarto will use Pandoc to convert to LaTeX, and then use LaTeX to do the PDF conversion (https://quarto.org/docs/output-formats/pdf-engine.html)
* Pandoc will directly call latex and do some processing when converting to PDF - Especially [calling `rsvg-convert` ](https://github.com/jgm/pandoc/blob/b263f383040618ff991912512ccd396b83d43da5/src/Text/Pandoc/PDF.hs#L194-L205) as described in #1144
That does not solve the issue, but it explains the different behavior. 

Quarto would need to mimic the Pandoc behavior for svg image - probably using a Lua filters could already do this but this is not yet built-in Quarto. Will do that soon.

**Comment by nluetts** [comment]
> @nluetts when you are using pandoc only, are you converting directly `--from markdown --to pdf` ?

@cderv yes, exactly. There is no intermediary `.tex` file produced (at least not actively by me).

Thank you for explaining the quarto behavior, now I understand the different outcome. Looking forward to the Lua filter :+1:

**Comment by fuhrmanator** [comment]
> I came across this same issue today. I just want to add (although I think it is already implied) that this issue is not related to just the book format. I was unable to render the following project because I included an SVG image.

As I mentioned above, SVG images in PDFs are possible by using the `pdf-engine: latexmk`. It's not perfect, however, since `rsvg-convert` doesn't grok advanced SVG (drawio). Here's the result I got for

```markdown
---
title: "Hello world"

format:
    pdf:
        pdf-engine: latexmk
---

This is my quarto doc. Here is an SVG image:

![](example-image.drawio.svg)
```

![image](https://user-images.githubusercontent.com/7606540/197195745-839ec334-2d04-498f-ae4c-4cdca7649c70.png)

**Comment by cderv** [comment]
> Quarto would need to mimic the Pandoc behavior for svg image - probably using a Lua filters could already do this but this is not yet built-in Quarto. Will do that soon (https://github.com/quarto-dev/quarto-cli/issues/2575)

Just a heads up that we did add support for `rsvg-convert` CLI tool to automatically convert SVG to PDF files when using the PDF output format.  This is available in 1.3 so using one of the latest pre-release.

**Comment by fuhrmanator** [comment]
@cderv thanks for supporting the SVG explicitly. When I have a chance, I will try to move my book project away from latexmk (I was only using it for SVG support).
